### ISSUE TITLE
Cookie `section_data_ids` isn't filled properly after restarting browser application

### ISSUE THREAD

**Issue by jasperzeinstra** [magento/magento2#14912]
### Preconditions
1. Magento 2.2.3 with sample data

### Steps to reproduce
1. Login with the test user account (roni_cost@example.com)
2. Notice the `Welcome Message`, the `Full Name` at the account dropdown and `Compare Products` link 
3. Now completely shut down your browser application (don't just close the browser window)
4. Restart your browser application
5. Go to the site and notice that your still logged in as expected, but that the `Welcome Message` is different, the `Full Name` at the account dropdown is missing and the link `Compare Products` is missing.

### Expected result
1. After a restart of your browser application (not window) the `Welcome Message` should contain the customer name, the `Full Name` should be shown at the account dropdown and the link `Compare Products` should be visible.

### Actual result
1. Screen recording: https://drive.google.com/file/d/18yCEJq67nCuaoh-vsUHXuLYKp68fLFgB/view

### Technical notes
* The missing data should be retreived from local storage `mage-cache-storage`
* The file `vendor/magento/module-customer/view/frontend/web/js/customer-data.js` fills / updates the local storage
* This file checks and creates cookie `section_data_ids`. The expire date of this cookie is `1969-12-31T23:59:59`
* After you restart your browser application this cookie isn't available anymore. Your local storage is empty. The cookie `PHPSESSID` still knows your loggedin. `customer-data.js` only triggers an update for `cart` and `directory-data` sections. Other sections are not going to be updated anymore and data such as `customer` and `compare-products` are missing in local storage.

### Screenshots before browser application restart (and after login)
![screen shot 2018-04-30 at 11 15 35](https://user-images.githubusercontent.com/9214557/39421269-93b33cc4-4c68-11e8-8a60-db39880b32a4.png)
![screen shot 2018-04-30 at 11 13 10](https://user-images.githubusercontent.com/9214557/39421270-98953d82-4c68-11e8-878c-b9cb345c8503.png)

### Screenshots after browser application restart
![screen shot 2018-04-30 at 11 14 55](https://user-images.githubusercontent.com/9214557/39421285-a42ce1ea-4c68-11e8-9e6f-83f23a4be908.png)
![screen shot 2018-04-30 at 11 14 28](https://user-images.githubusercontent.com/9214557/39421288-a769ebe6-4c68-11e8-9a0c-38554dd28cfe.png)

**Comment by receter** [comment]
We have the same issue. Magento ver. 2.2.2

Is there any quick-fix/workaround to address this problem?

**Comment by ghost** [comment]
HI @jasperzeinstra thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by jfmontanaro** [comment]
Can confirm this is also an issue on 2.2.5.  

I have found a possible workaround, however.  There are 3 places where `customer-data.js` sets that cookie.  In 2.2.5 it happens on lines 112, 173, and 353.  If you change `$.cookieStorage.set` on those lines to `$.cookieStorage.setExpires(n).set`, where `n` is the number of days for which you would like the cookie to remain valid, the issue disappears.  I used n=7 since we retain customer sessions for 7 days.

I have not tested this extensively so I don't know whether it might cause other problems elsewhere.  Use at your own risk.

**Comment by pmclain** [comment]
Looks to be fixed in 2.3-develop https://github.com/magento/magento2/commit/b13d3b48b779f1d5b67b07d9d6e0d69daf465d5b

**Comment by vuongggggg** [comment]
> Looks to be fixed in 2.3-develop [b13d3b4](https://github.com/magento/magento2/commit/b13d3b48b779f1d5b67b07d9d6e0d69daf465d5b)

That "fix" already being in the source code.  I mean this problem still not be solved because I still got it and this ticket still opening for answer. Any update?

**Comment by magento-engcom-team** [comment]
Hi @jasperzeinstra. 

Thank you for your report and collaboration!

The related internal Jira ticket `MAGETWO-93625` was closed as `non-reproducible` in `2.2.x`.
It means that Magento team either unable to reproduce this issue using provided _Steps to Reproduce_ from the **_Description section_** on clean or the issue has been already fixed in the scope of other tasks.

But if you still run into this problem please update or provide additional information/steps/preconditions in the **_Description section_** and reopen this issue.

**Comment by trunglv** [comment]
Due to customer data on Local Storage is not loaded. So my solution is to check whether customer data exists on the Local Storage or does not. Then if it doesn't, we will make it expired. So it will be loaded again from the server again.

Here a quick fix for this issue below. But I think the Magento team should check and fix it an issue.

Create a mixin

```
var config = {
 config: {
     mixins: {
         'Magento_Customer/js/customer-data': {
             
             'FixBugs_CustomerLocalStorageIssue/js/customer-data-mixin': true
             
         }
     }
 }
};   

```

```
/**
 * trunglv - Fixing bugs
 * 
 */
define([
    'jquery',
    'mage/utils/wrapper'
], function ($, wrapper) {
    'use strict';

    var mixin = {
        /**
         * Check if persistent section is expired due to lifetime.
         *
         * @param {Function} originFn - Original method.
         * @return {Array}
         */
        getExpiredSectionNames: function (originFn) {
            var expiredSectionNames = originFn(),
                storage = $.initNamespaceStorage('mage-cache-storage').localStorage,            
                sectionData;
            
            /* to add more if you want. It temporary fix for customer data on local storage */
            var sectionNames = ['customer'];
            
            _.each(sectionNames, function (sectionName) {
                
                sectionData = storage.get(sectionName);
                
                if(typeof sectionData === 'undefined'){
                        expiredSectionNames.push(sectionName);
                }
                
            });
            return expiredSectionNames;
        }
    };

    /**
     * Override default customer-data.getExpiredSectionNames().
     */
    return function (target) {
        return wrapper.extend(target, mixin);
    };
});
```

--- Another solution, but need to fix core  -- 

Add section "customer" into expiredSectionNames 
<type name="Magento\Customer\Block\CustomerData">
        <arguments>
            <argument name="expirableSectionNames" xsi:type="array">
                <item name="customer" xsi:type="string">customer</item>
            </argument>
        </arguments>
    </type>

In JS file : Magento_Customer/js/customer-data

```
getExpiredSectionNames: function () {
            var expiredSectionNames = [],
                cookieSectionTimestamps = $.cookieStorage.get('section_data_ids') || {},
                sectionLifetime = options.expirableSectionLifetime * 60,
                currentTimestamp = Math.floor(Date.now() / 1000),
                sectionData;

            // process sections that can expire due to lifetime constraints
            _.each(options.expirableSectionNames, function (sectionName) {
                sectionData = storage.get(sectionName);

                if (typeof sectionData === 'object' && sectionData['data_id'] + sectionLifetime <= currentTimestamp) {
                    expiredSectionNames.push(sectionName);
                }else{
                    /* Here is my fixing*/
                    if(typeof sectionData == 'undefined'){
                        expiredSectionNames.push(sectionName);
                    }
                }
            });

            // process sections that can expire due to storage information inconsistency
            _.each(cookieSectionTimestamps, function (cookieSectionTimestamp, sectionName) {
                sectionData = storage.get(sectionName);

                if (typeof sectionData === 'undefined' ||
                    typeof sectionData === 'object' &&
                    cookieSectionTimestamp != sectionData['data_id'] //eslint-disable-line
                ) {
                    expiredSectionNames.push(sectionName);
                }
            });

            return _.uniq(expiredSectionNames);
        },
```

**Comment by jorge-rollpix** [comment]
> Due to customer data in Local Storage is not loaded. So my solution is to check whether customer data exists in the Local Storage or does not. Then if it doesn't, we will make it expired. So it will be loaded again from the server.
> 
> Here a quick fix for this issue below. But I think the Magento team should check and fix it an issue.
> 
> Create a mixin
> 
> ```
> var config = {
>  config: {
>      mixins: {
>          'Magento_Customer/js/customer-data': {
>              
>              'FixBugs_CustomerLocalStorageIssue/js/customer-data-mixin': true
>              
>          }
>      }
>  }
> };   
> ```
> 
> ```
> /**
>  * trunglv - Fixing bugs
>  * 
>  */
> define([
>     'jquery',
>     'mage/utils/wrapper'
> ], function ($, wrapper) {
>     'use strict';
> 
>     var mixin = {
>         /**
>          * @param {Function} originFn - Original method.
>          * @return {Array}
>          */
>         getExpiredSectionNames: function (originFn) {
>             var expiredSectionNames = originFn(),
>                 storage = $.initNamespaceStorage('mage-cache-storage').localStorage,            
>                 sectionData;
>             
>             /* to add more if you want. It temporary fix for customer data on local storage */
>             var sectionNames = ['customer'];
>             
>             _.each(sectionNames, function (sectionName) {
>                 
>                 sectionData = storage.get(sectionName);
>                 
>                 if(typeof sectionData === 'undefined'){
>                         expiredSectionNames.push(sectionName);
>                 }
>                 
>             });
>             return expiredSectionNames;
>         }
>     };
> 
>     /**
>      * Override default customer-data.getExpiredSectionNames().
>      */
>     return function (target) {
>         return wrapper.extend(target, mixin);
>     };
> });
> ```
> 
> --- Another solution, but need to fix core --
> 
> Add section "customer" into expiredSectionNames
> 
> ```
> <type name="Magento\Customer\Block\CustomerData">
>         <arguments>
>             <argument name="expirableSectionNames" xsi:type="array">
>                 <item name="customer" xsi:type="string">customer</item>
>             </argument>
>         </arguments>
>     </type>
> ```
> 
> In JS file : Magento_Customer/js/customer-data
> 
> ```
> getExpiredSectionNames: function () {
>             var expiredSectionNames = [],
>                 cookieSectionTimestamps = $.cookieStorage.get('section_data_ids') || {},
>                 sectionLifetime = options.expirableSectionLifetime * 60,
>                 currentTimestamp = Math.floor(Date.now() / 1000),
>                 sectionData;
> 
>             // process sections that can expire due to lifetime constraints
>             _.each(options.expirableSectionNames, function (sectionName) {
>                 sectionData = storage.get(sectionName);
> 
>                 if (typeof sectionData === 'object' && sectionData['data_id'] + sectionLifetime <= currentTimestamp) {
>                     expiredSectionNames.push(sectionName);
>                 }else{
>                     /* Here is my fixing*/
>                     if(typeof sectionData == 'undefined'){
>                         expiredSectionNames.push(sectionName);
>                     }
>                 }
>             });
> 
>             // process sections that can expire due to storage information inconsistency
>             _.each(cookieSectionTimestamps, function (cookieSectionTimestamp, sectionName) {
>                 sectionData = storage.get(sectionName);
> 
>                 if (typeof sectionData === 'undefined' ||
>                     typeof sectionData === 'object' &&
>                     cookieSectionTimestamp != sectionData['data_id'] //eslint-disable-line
>                 ) {
>                     expiredSectionNames.push(sectionName);
>                 }
>             });
> 
>             return _.uniq(expiredSectionNames);
>         },
> ```

Could you put the full path of each file please?

**Comment by engcom-Bravo** [comment]
Hello @jasperzeinstra 

Thank you for your report

This issue is already fixed in the latest 2.4-develop

When the browser is **Closed** and **Restarted** the Compare products information contain compared products and Welcome message displays Customers name
![After restart](https://user-images.githubusercontent.com/51680745/96453624-686c1080-1223-11eb-9036-e09265f66175.gif)

If you are still experiencing issues on the latest code base, please feel free to comment, reopen or create a new ticket according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines)
